### ISSUE TITLE
CI failure: "Missing digest"

### ISSUE THREAD

**Issue by BetsyMcPhail** [RobotLocomotion/drake#21255]
On 4/2/24, buildcops began noticing CI jobs failing with errors like:

`[10:15:14 AM]  ERROR: /media/ephemeral0/ubuntu/workspace/linux-jammy-gcc-bazel-continuous-debug/src/math/BUILD.bazel:427:20: Compiling math/test/discrete_algebraic_riccati_equation_test.cc failed: unable to finalize action: Missing digest: 6f5ad5446e49451a55dddee73681d3f44103b4aea9c85b308a9761effe9e3693/34778 for bazel-out/k8-dbg/bin/math/_objs/discrete_algebraic_riccati_equation_test/discrete_algebraic_riccati_equation_test.pic.d`

The build cache key was incremented in https://github.com/RobotLocomotion/drake-ci/pull/276 as an attempted solution, but the failure continues to sporadically appear. 

See [slack conversation](https://drakedevelopers.slack.com/archives/C270MN28G/p1712073402141219) for initial discussion.

**Comment by BetsyMcPhail** [comment]
4/15: https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/495/

**Comment by SeanCurtis-TRI** [comment]
4/18 

https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/514/

**Comment by xuchenhan-tri** [comment]
4/24
[https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/537/](https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/537/)

**Comment by BetsyMcPhail** [comment]
5/20
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-debug/283/ and https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-debug/284/

**Comment by mwoehlke-kitware** [comment]
5/28
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/622/

**Comment by BetsyMcPhail** [comment]
6/4 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-debug/320/

**Comment by xuchenhan-tri** [comment]
https://drake-jenkins.csail.mit.edu/view/Production/job/linux-jammy-clang-bazel-continuous-debug/351
https://drake-jenkins.csail.mit.edu/view/Production/job/linux-jammy-clang-bazel-continuous-debug/352

**Comment by BetsyMcPhail** [comment]
8/5 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/787/

**Comment by mwoehlke-kitware** [comment]
8/6 https://drake-jenkins.csail.mit.edu/view/Production/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/796/

**Comment by BetsyMcPhail** [comment]
9/10
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/867/
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/868/

**Comment by nicolecheetham** [comment]
10/9 https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/960/

**Comment by williamjallen** [comment]
10/16

https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/984/consoleFull

**Comment by nicolecheetham** [comment]
11/6 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1041/consoleFull

**Comment by BetsyMcPhail** [comment]
12/12 cont.

https://drake-jenkins.csail.mit.edu/job/linux-jammy-gcc-bazel-continuous-debug/2239/
https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1067/
https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-debug/745/

**Comment by BetsyMcPhail** [comment]
1/6: https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1140

**Comment by ggould-tri** [comment]
https://drake-jenkins.csail.mit.edu/view/Production/job/linux-jammy-gcc-bazel-continuous-debug/2330/consoleFull again x4

**Comment by jwnimmer-tri** [comment]
For the record, I believe there is not much we can do to fix this.  The rate is elevated this week because I'm pushing dozens of PRs that invalidate the entire build cache.  Once that quiets down, the error frequency should drop back down to a more nominal rate.

**Comment by jwnimmer-tri** [comment]
FYI If it's a PR flake that's the only thing preventing merging, you can ping me or sherm1 to force it in anyway, you don't need to re-trigger.

**Comment by BetsyMcPhail** [comment]
1/14 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1194

**Comment by williamjallen** [comment]
1/22 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1232/consoleFull

**Comment by jwnimmer-tri** [comment]
1/28:
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-experimental-leak-sanitizer/7054/consoleFull
https://drake-jenkins.csail.mit.edu/job/linux-jammy-gcc-bazel-experimental-debug/11393/consoleFull

**Comment by BetsyMcPhail** [comment]
2/4:
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1300/
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-debug/977

**Comment by BetsyMcPhail** [comment]
2/12 https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1319

**Comment by BetsyMcPhail** [comment]
2/14 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-debug/1010/

**Comment by tyler-yankee** [comment]
3/3: https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-debug/1041/

**Comment by tyler-yankee** [comment]
3/3: https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1360/

**Comment by tyler-yankee** [comment]
3/4: https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-release/1464

**Comment by xuchenhan-tri** [comment]
3/4: [https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1373](https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1373)

**Comment by xuchenhan-tri** [comment]
3/5: https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1379

**Comment by xuchenhan-tri** [comment]
3/6: https://drake-jenkins.csail.mit.edu/view/Production/job/linux-jammy-gcc-bazel-continuous-debug/2555/

**Comment by BetsyMcPhail** [comment]
3/17 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1424

**Comment by BetsyMcPhail** [comment]
3/20 https://drake-jenkins.csail.mit.edu/job/linux-jammy-gcc-bazel-continuous-everything-release/1113

**Comment by Aiden2244** [comment]
3/27 https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1457/

**Comment by tyler-yankee** [comment]
4/7: https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1494/

**Comment by BetsyMcPhail** [comment]
4/8 https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-clang-bazel-continuous-everything-address-sanitizer/1501/

**Comment by BetsyMcPhail** [comment]
4/9 https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-gcc-bazel-continuous-debug/2672/

**Comment by SeanCurtis-TRI** [comment]
4/15 https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-gcc-bazel-continuous-debug/2689/

**Comment by BetsyMcPhail** [comment]
4/29 https://drake-jenkins.csail.mit.edu/view/Continuous%20Production/job/linux-jammy-gcc-bazel-continuous-everything-release/1227/

**Comment by BetsyMcPhail** [comment]
5/5
https://drake-jenkins.csail.mit.edu/job/linux-jammy-gcc-bazel-continuous-debug/2744/
https://drake-jenkins.csail.mit.edu/job/linux-jammy-gcc-bazel-continuous-debug/2745
https://drake-jenkins.csail.mit.edu/job/linux-jammy-clang-bazel-continuous-leak-sanitizer/1573

**Comment by jwnimmer-tri** [comment]
Maybe we should try adding `--experimental_remote_cache_lease_extension=yes` to see it that helps?

We momentarily had it for other reasons (https://github.com/RobotLocomotion/drake-ci/pull/324 -- remote download minimal) but ripped it out again when it didn't fix that problem.  Maybe it fixes this problem?

**Comment by tyler-yankee** [comment]
Sounds like it's at least worth a shot. If that doesn't work, there might be also `--noexperimental_inmemory_dotd_files` and/or `--noexperimental_inmemory_jdeps_files`, though this might slow the builds down as it would write the `.d` / `.jdeps` files to disk instead of keeping them in memory (https://github.com/bazelbuild/bazel/issues/22854).

**Comment by ggould-tri** [comment]
https://drake-jenkins.csail.mit.edu/view/Production/job/linux-jammy-gcc-bazel-continuous-debug/2755/

**Comment by tyler-yankee** [comment]
I just doubled the size of the remote cache disk (1 TB -> 2 TB). FWIW, as I'm writing it's using 713 GB. 

Since this issue seems to occur in bunches (i.e. when multiple builds run at once), perhaps they're all trying to add files to the remote cache and "older" files that are actually still by other builds are getting evicted.

**Comment by tyler-yankee** [comment]
I suppose another thing to look into could be ([docs](https://bazel.build/remote/caching)):

> Starting with Bazel 7.4, you can use `--experimental_disk_cache_gc_max_size` and `--experimental_disk_cache_gc_max_age` to set a maximum size for the disk cache or for the age of individual cache entries. Bazel will automatically garbage collect the disk cache while idling between builds; the idle timer can be set with `--experimental_disk_cache_gc_idle_delay` (defaulting to 5 minutes).

**Comment by ggould-tri** [comment]
https://drake-jenkins.csail.mit.edu/job/linux-jammy-gcc-bazel-experimental-debug/12690/consoleFull

**Comment by jwnimmer-tri** [comment]
> I suppose another thing to look into could be ... experimental_disk_cache_gc_max_size ...

That's for the local disk cache, not the remote cache.

**Comment by tyler-yankee** [comment]
We did a bunch in the above PR, plus doubled the size of the disk for the cache. We haven't seen the error come up in a few days, so if we see it again, we'll turn on the new verbose logging option and start debugging.

**Comment by BetsyMcPhail** [comment]
7/30 https://drake-jenkins.csail.mit.edu/job/linux-noble-clang-bazel-continuous-everything-address-sanitizer/88/

**Comment by Aiden2244** [comment]
12/1 https://drake-jenkins.csail.mit.edu/job/linux-noble-gcc-bazel-continuous-debug/617/

**Comment by BetsyMcPhail** [comment]
12/9 https://drake-jenkins.csail.mit.edu/job/linux-noble-clang-bazel-continuous-everything-address-sanitizer/640/

**Comment by tyler-yankee** [comment]
We have been seeing more frequent occurrences including many that aren't documented here.

https://bazel.build/remote/caching#delete-remote-cache pretty much outlines our exact problem, but doesn't seem to offer any solutions.

Some ideas:

* RobotLocomotion/drake-ci#326 added the `experimental_remote_cache_lease_extension` flag, but per https://bazel.build/reference/command-line-reference#common_options-flag--experimental_remote_cache_lease_extension the frequency with which it pings the cache server to update "in use" timestamps is 3 hours. We could try playing with this time interval via `--experimental_remote_cache_ttl`.
* https://bazel.build/reference/command-line-reference#build-flag--experimental_remote_cache_eviction_retries

#23708 will also help this problem some by reducing churn for back to back jobs, but I think there should be something in place to handle this issue on its own as well.

**Comment by tyler-yankee** [comment]
With the above CI change merged, next steps are to monitor for any issues, and it'll probably take a couple weeks of not seeing any `Missing digest` errors before we can call this issue closed. If the frequency decreases but it still occurs, perhaps we just tweak the config value we just set.

**Comment by tyler-yankee** [comment]
We immediately had a continuous failure with `Missing digest`: https://drake-jenkins.csail.mit.edu/job/linux-noble-clang-bazel-continuous-everything-address-sanitizer/708/

But this is possibly/probably due to leftover blobs in the cache that hadn't had the new flag applied yet? I'd like to see a failure e.g. tomorrow before we resolve that the above PR doesn't fix.

**Comment by BetsyMcPhail** [comment]
Failure after the latest MR: https://drake-jenkins.csail.mit.edu/job/linux-noble-gcc-bazel-continuous-debug/749/

**Comment by tyler-yankee** [comment]
Enough evidence from this and some builds that weren't posted here, that this flag isn't a fix either.

**Comment by jwnimmer-tri** [comment]
Once we're using Bazel 9.1, we should try using the new feature `--rewind_lost_inputs` in CI.

**Comment by jwnimmer-tri** [comment]
We using Bazel 9.1 now.

**Comment by tyler-yankee** [comment]
We haven't seen this issue in several months. Please reopen if it occurs again, but otherwise I'll close and track the `--rewind_lost_inputs` mentioned in https://github.com/RobotLocomotion/drake/issues/21255#issuecomment-4225523576 in another issue.
